doc/userguide/rules/rules-internals.rst

```diff
@@ -0,0 +1,239 @@
+this state, the signatures with those keywords may be already evaluated, even if
+they have a lower priority than an http body inspecting signature.
+
+Relatedly, a rule with two keywords matching at two different progress stages may
+be evaluated against two different packets.
+
+Implications
+------------
+
+Action precedence and interaction with ``ip-only`` rules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To illustrate what may be counter-intuitive implications of how inspection
+steps, action prioritization and rule keywords interact and affect the engine
+behavior, we will use a real case example. Consider these three rules:
+
+.. container:: example-rule
+
+    pass tcp 0.0.0.0/0 any <> 0.0.0.0/0 443 (msg:"Allow TCP in port 443"; flow:
+    not_established; sid:1; rev:1;)
+
+.. container:: example-rule
+
+    pass tcp 0.0.0.0/0 any <> 0.0.0.0/0 80 (msg:"Allow TCP in port 80"; flow:
+    not_established; sid:2; rev:1;)
+
+.. container:: example-rule
+
+    drop ip 0.0.0.0/0 any -> 0.0.0.0/0 any (msg:"No outbound internet access
+    from host"; sid:3; rev:1;)
+
+The first two are signatures that analyze individual packets and match only if
+the flow has not been established (``flow:not_established``): the rules grant
+``PASS`` to the matched packet - but not to its flow.
+
+The third signature is considered ``ip-only``. This means it will be evaluated
+for the *first* packet in both directions of a flow, in addition to rules 1 and
+2. By extension, **the other packets in the same flow will not be evaluated
+against this rule.**
+
+With an action order configuration that prioritizes ``PASS`` over ``DROP``, this
+means that rules 1 and 2 will have a higher internal priority over rule 3,
+therefore nullifying the ``DROP`` outcome. The result: a flow for outbound
+internet traffic from the host, expected to be dropped, wouldn't be.
+
+If the expected behavior with those three signatures was to allow traffic on
+ports 80 and 443 only, while dropping everything else, the simplest way to
+achieve this would be to remove the ``flow:not_established`` portion from rules
+sid:1 and sid:2. This ensures that the ``PASS`` action would be applied to the
+whole flow following the match on the first packet and that all other traffic
+would be dropped.
+
+Following that, all three rules will be evaluated on the same step, and if a
+flow isn't flagged with ``pass``, it will be dropped with the third rule.
+
+.. Tip::
+   A more straightforward way to achieve that in Suricata 8 is using the firewall
+   more. See :doc:`../firewall/firewall-design`.
+
+Alerts not seen
+~~~~~~~~~~~~~~~
+
+Another aspect of rule prioritization combined with the alerts queue size, is
+that in corner case scenarios, if a packet matches against too many rules,
+signatures with lower priority could be discarded from the ``PacketAlert`` queue
+(see the section on :ref:`alert queue overflow impact` for more).
```

Comment:
Can we know if there was a alert overflow in a packet ?
I do not think so, and I think we should have a ticket for this improvement...

Reply:
Not per packet, but we do have stats counters: https://docs.suricata.io/en/latest/configuration/suricata-yaml.html#discarded-and-suppressed-alerts-stats
```
detect.alert_queue_overflow
detect.alerts_suppressed
```
I can mention these counters here. Do you suggest we have a way to know for a particular packet if there was an alert overflow?

Reply:
> I can mention these counters here

Good idea :-)

> Do you suggest we have a way to know for a particular packet if there was an alert overflow?

I did not know we had the stats, so I think it is ok even if it could be better :-)